user/overriding-packages.rst

```diff
@@ -0,0 +1,283 @@
+This page offers tips for avoiding most issues.
+
+.. contents:: Table of Contents
+    :depth: 3
+
+
+How to avoid most issues
+------------------------
+
+These are good practices to avoid most issues.
+If the advice is too restrictive then see the known issue descriptions for more complex advice.
+
+Use isolated workspaces
+***********************
+
+If you are building a workspace and suspect you may overriding a package from it in the future, then use an isolated workspace.
+This avoids include directory search order issues.
+
+Override every package that depends on the one you want to override
+*******************************************************************
+
+A **leaf package** is one that has no other packages that depend on it.
+Overriding a non-leaf package can be problematic.
+Packages in the underlay were built against the underlay version of the package, but they will be expected to run with the overlay version.
+If their build process stores some information about the non-leaf package, such as an expected ABI, then undefined behavior can happen at runtime.
+
+Problems caused by packages remembering information at build time can be avoided by overriding every package that directly or indirectly depends on the one you actually want to override.
+The group of overridden packages must span all underlays.
+
+Say there are 3 workspaces (**A**, **B** and **C**) where **C** overlays **B** which overlays **A**.
+**A** contains packages ``foo`` and ``baz`` where ``baz`` depends on ``foo``.
+**B** contains packages ``ping`` that depends on ``foo`` and  ``pong`` that depends on ``baz``.
+**C** is the workspace being built.
+If you want to override ``foo`` then you should also override ``baz``, ``ping``, and ``pong``.
+
+Make sure overridden Python packages do not change entry point specifications
+*****************************************************************************
+
+Python packages may have `entry point specifications <https://packaging.python.org/en/latest/specifications/entry-points/>`_ in a ``setup.py`` or ``setup.cfg`` file.
+Make sure the package in the overlay has identical specifications to the version in the underlay, or only adds new ones.
+If any specification has been changed or removed then it may not be possible to override this package.
+
+How to make it easier for your users override
+---------------------------------------------
+
+This section has advice for package authors about how to make easier for your users to use your package and override it or other pacakges.
+
+Install headers to a unique include directory
+*********************************************
+
+Install your package's headers to a unique folder rather than a shared folder.
+Say you're the author of a package ``foo``, and it has a header meant to be included like ``#include <foo/foo.hpp>``.
+Instead of installing the header to ``<prefix>/include/foo/foo.hpp``, install it to ``<prefix>/include/foo/foo/foo.hpp``.
+
+Here's an example.
```

Comment:
Aren't the two lines above also an example? Perhaps re-order this section.

Reply:
Reworded in fbf7704